// gueco/interfaz/usuario/fachada/interno/servicio_usuario.js
/*! \package interfaz.usuario.fachada
 */

var especificador_objeto =
    require('gueco/dominio/comun/especificador_objeto')


/*! autenticar
 * \brief Autentica
 * \param identificador Nombre de usuario
 * \param clave Clave
 * \param retrollamada función (error, usuario) {}
 *
 * restricciones:
 *     contexto fachada_servicio_usuario::autenticar(
 *         identificador: texto, clave: texto, retrollamada: función)
 *
 *         pre: retrollamada._es(función (error: Error, usuario: usuario))
 */
fachada_servicio_usuario.prototype.autenticar =
    function(identificador, clave, retrollamada) {

    this.servicio_usuario.autenticar(
        {
            identificador: identificador,
            clave: clave
        },
        function(error, usuario) {

        retrollamada(error, error ? null : usuario)
    })
}

fachada_servicio_usuario.prototype.escenarios = function(usuario,
                                                         retrollamada) {
    this.servicio_usuario.escenarios(usuario, function(error, escenarios) {
        var otd = null
        if (!error) {
            otd = {}
            for (var i in escenarios) {
                var escenario = escenarios[i]
                otd[escenario.id] = escenario
                delete escenario.id
            }
        }
        retrollamada(error, otd)
    })
}

fachada_servicio_usuario.prototype.escenario = function(id, retrollamada) {
    this.repositorio_escenario.uno({id: id}, retrollamada)
}

fachada_servicio_usuario.prototype.almacenar_escenarios =
    function(usuario, escenarios, retrollamada) {

    this.servicio_usuario.almacenar_escenarios(usuario, escenarios,
                                               retrollamada)
}

fachada_servicio_usuario.prototype.borrar_escenario =
    function(id, cuerpo, usuario, retrollamada) {

    this.repositorio_escenario.borrar({ id: id })

    var i = usuario.escenarios.indexOf(id)
    if (-1 != i)
       usuario.escenarios.splice(i, 1)

    this.servicio_usuario.repositorio_usuario.almacenar(usuario)

    retrollamada(null)
}

function fachada_servicio_usuario () {}

module.exports = fachada_servicio_usuario
